runtime/query.go

```diff
@@ -45,362 +46,294 @@ func (*defaultQueryParser) Parse(msg proto.Message, values url.Values, filter *u
+		}
+		return protoreflect.ValueOfInt64(v), nil
+	case protoreflect.Uint32Kind, protoreflect.Fixed32Kind:
+		v, err := strconv.ParseUint(value, 10, 32)
+		if err != nil {
+			return protoreflect.Value{}, fmt.Errorf("parsing unsigned integer: %w", err)
+		}
+		return protoreflect.ValueOfUint32(uint32(v)), nil
+	case protoreflect.Uint64Kind, protoreflect.Fixed64Kind:
+		v, err := strconv.ParseUint(value, 10, 64)
+		if err != nil {
+			return protoreflect.Value{}, fmt.Errorf("parsing unsigned integer: %w", err)
+		}
+		return protoreflect.ValueOfUint64(v), nil
+	case protoreflect.FloatKind:
+		v, err := strconv.ParseFloat(value, 32)
+		if err != nil {
+			return protoreflect.Value{}, fmt.Errorf("parsing float: %w", err)
+		}
+		return protoreflect.ValueOfFloat32(float32(v)), nil
+	case protoreflect.DoubleKind:
+		v, err := strconv.ParseFloat(value, 64)
+		if err != nil {
+			return protoreflect.Value{}, fmt.Errorf("parsing float: %w", err)
+		}
+		return protoreflect.ValueOfFloat64(v), nil
+	case protoreflect.StringKind:
+		return protoreflect.ValueOfString(value), nil
+	case protoreflect.BytesKind:
+		v, err := base64.StdEncoding.DecodeString(value)
+		if err != nil {
+			return protoreflect.Value{}, fmt.Errorf("parsing base64 bytes: %w", err)
 		}
-		f.Index(i).Set(result[0].Convert(f.Index(i).Type()))
+		return protoreflect.ValueOfBytes(v), nil
+	case protoreflect.MessageKind, protoreflect.GroupKind:
+		return parseMessage(msgValue.Get(fieldDesc).Message(), value)
+	default:
+		panic(fmt.Sprintf("unknown field kind: %v", fieldDesc.Kind()))
 	}
-	return nil
 }
 
-func populateField(f reflect.Value, value string, props *proto.Properties) error {
-	i := f.Addr().Interface()
+func parseMessage(msgValue protoreflect.Message, value string) (protoreflect.Value, error) {
+	msgDesc := msgValue.Descriptor()
 
 	// Handle protobuf well known types
-	var name string
-	switch m := i.(type) {
-	case interface{ XXX_WellKnownType() string }:
-		name = m.XXX_WellKnownType()
-	case proto.Message:
-		const wktPrefix = "google.protobuf."
-		if fullName := proto.MessageName(m); strings.HasPrefix(fullName, wktPrefix) {
-			name = fullName[len(wktPrefix):]
-		}
+	if msgDesc.FullName().Parent() != "google.protobuf" {
+		return protoreflect.Value{}, fmt.Errorf("only well known type messages are supported")
 	}
-	switch name {
+
+	msg := msgValue.New()
+	switch msgDesc.Name() {
 	case "Timestamp":
 		if value == "null" {
-			f.FieldByName("Seconds").SetInt(0)
-			f.FieldByName("Nanos").SetInt(0)
-			return nil
+			break
 		}
 
 		t, err := time.Parse(time.RFC3339Nano, value)
 		if err != nil {
-			return fmt.Errorf("bad Timestamp: %v", err)
+			return protoreflect.Value{}, fmt.Errorf("parsing Timestamp: %w", err)
 		}
-		f.FieldByName("Seconds").SetInt(int64(t.Unix()))
-		f.FieldByName("Nanos").SetInt(int64(t.Nanosecond()))
-		return nil
+		msg.Set(msgDesc.Fields().ByName("seconds"), protoreflect.ValueOfInt64(t.Unix()))
```

Comment:
Is there no other way to do this, ie convert message to a proper `Timestamp` and then set the fields?

Reply:
Can either use [`ptypes.TimestampProto`](https://pkg.go.dev/github.com/golang/protobuf/ptypes?tab=doc#TimestampProto) or wait until https://golang.org/cl/225298 is submitted (I've been distracted by many things lately).

Reply:
Changed to instantiate all messages explicitly, this avoids the need to pass around `msgValue` which was a pain point. The new code looks much nicer!